src/audio/SDL_audiocvt.c

```diff
@@ -869,6 +874,108 @@ bool SDL_PutAudioStreamData(SDL_AudioStream *stream, const void *buf, int len)
+                *(dst++) = srcs[channel][frame]; \
+            } \
+        } \
+    }
+
+GENERIC_INTERLEAVE_FUNCTION(8)
+GENERIC_INTERLEAVE_FUNCTION(16)
+GENERIC_INTERLEAVE_FUNCTION(32)
+//GENERIC_INTERLEAVE_FUNCTION(64)   (we don't have any 64-bit audio data types at the moment.)
+#undef GENERIC_INTERLEAVE_FUNCTION
+
+static void InterleaveAudioChannels(void *output, const void * const *channel_buffers, int num_samples, const SDL_AudioSpec *spec)
+{
+    const int channels = spec->channels;
+
+    // !!! FIXME: it would be possible to do this really well in SIMD for stereo data, using unpack (intel) or zip (arm) instructions, etc.
+
+    switch (SDL_AUDIO_BITSIZE(spec->format)) {
+        case 8: InterleaveAudioChannelsGeneric8(output, channel_buffers, channels, num_samples); break;
+        case 16: InterleaveAudioChannelsGeneric16(output, channel_buffers, channels, num_samples); break;
+        case 32: InterleaveAudioChannelsGeneric32(output, channel_buffers, channels, num_samples); break;
+        //case 64: InterleaveAudioChannelsGeneric64(output, channel_buffers, channels, num_samples); break;  (we don't have any 64-bit audio data types at the moment.)
+        default: SDL_assert(!"Missing needed generic audio interleave function!"); SDL_memset(output, 0, SDL_AUDIO_FRAMESIZE(*spec) * num_samples); break;
+    }
+}
+
+bool SDL_PutAudioStreamSeparatedData(SDL_AudioStream *stream, const void * const *channel_buffers, int num_samples)
+{
+    if (!stream) {
+        return SDL_InvalidParamError("stream");
+    } else if (!channel_buffers) {
+        return SDL_InvalidParamError("channel_buffers");
+    } else if (num_samples < 0) {
+        return SDL_InvalidParamError("num_samples");
+    } else if (num_samples == 0) {
+        return true; // nothing to do.
+    }
+
+    // we do the interleaving up front without the lock held, so the audio device doesn't starve while we work.
+    //  but we _do_ need to know the current input spec.
+    SDL_AudioSpec spec;
+    int chmap_copy[SDL_MAX_CHANNELMAP_CHANNELS];
+    int *chmap = NULL;
+    SDL_LockMutex(stream->lock);
+    if (!CheckAudioStreamIsFullySetup(stream)) {
+        SDL_UnlockMutex(stream->lock);
+        return false;
+    }
+    SDL_copyp(&spec, &stream->src_spec);
+    if (stream->src_chmap) {
+        chmap = chmap_copy;
+        SDL_memcpy(chmap, stream->src_chmap, sizeof (*chmap) * spec.channels);
+    }
+    SDL_UnlockMutex(stream->lock);
+
+    if (spec.channels == 1) {  // nothing to interleave, just use the usual function.
+        return SDL_PutAudioStreamData(stream, channel_buffers[0], SDL_AUDIO_FRAMESIZE(spec) * num_samples);
+    }
+
+    bool retval = false;
+
+    const int len = SDL_AUDIO_FRAMESIZE(spec) * num_samples;
+    #if DEBUG_AUDIOSTREAM
+    SDL_Log("AUDIOSTREAM: wants to put %d bytes of separated data", len);
+    #endif
+
+    // Is the data small enough to just interleave it on the stack and put it through the normal interface?
```

Comment:
Any reason not to use SDL_small_alloc()?

Reply:
Because it only allows 128 bytes of stack, and we'd need to SDL_small_free it afterwards (or break the abstraction). Here, we want to keep the malloc'd version, which gets reused directly as the backing for an SDL_AudioTrack.